crates/conjure_core/src/solver/adaptors/sat_adaptor.rs

```diff
@@ -0,0 +1,333 @@
+use std::any::type_name;
+use std::fmt::format;
+use std::iter::Inspect;
+use std::ptr::null;
+use std::vec;
+
+use clap::error;
+use minion_rs::ast::Model;
+use rustsat::encodings::am1::Def;
+use rustsat::solvers::{Solve, SolverResult};
+use rustsat::types::{Var as satVar, Clause, Lit}; 
+use sat_rs::sat_tree::{self, conv_to_clause, conv_to_formula};
+use std::collections::{HashMap, VecDeque}; 
+use crate::context::Context;
+
+use rustsat_minisat::core::Minisat;
+
+use crate::ast::{Expression, Name};
+use crate::metadata::Metadata;
+use crate::solver::{self, SearchStatus, SolveSuccess, SolverCallback, SolverFamily, SolverMutCallback};
+use crate::stats::SolverStats;
+use crate::{ast as conjure_ast, model, Model as ConjureModel};
+
+use super::super::model_modifier::NotModifiable;
+use super::super::private;
+use super::super::SearchComplete::*;
+use super::super::SearchIncomplete::*;
+use super::super::SearchStatus::*;
+use super::super::SolverAdaptor;
+use super::super::SolverError;
+use super::super::SolverError::*;
+use super::super::SolverError::*;
+
+use rustsat::instances::SatInstance;
+use thiserror::Error;
+
+/// A [SolverAdaptor] for interacting with the SatSolver generic and the types thereof.
+
+pub struct SAT {
+    __non_constructable: private::Internal,
+    model_inst: Option<SatInstance>,
+    var_map: Option<HashMap<i32, satVar>>,         // Mapping ConjureModel -> Minisat variables
+    reverse_var_map: Option<HashMap<satVar, i32>>, // Reverse mapping Minisat -> ConjureModel variables
+    solver_inst: Minisat,
+    incremental_clauses: VecDeque<Clause>,
+}
+
+impl private::Sealed for SAT {}
+
+impl Default for SAT {
+    fn default() -> Self {
+        SAT {
+            __non_constructable: private::Internal,
+            model_inst: None,
+            var_map: None,
+            reverse_var_map: None,
+            solver_inst: Minisat::default(),
+            incremental_clauses: VecDeque::new(),
+        }
+    }
+}
+
+impl SAT {
+    pub fn new(model: ConjureModel) -> Self {
+        let model_to_use: Option<SatInstance> = Some(SatInstance::new());
+        SAT {
+            __non_constructable: private::Internal,
+            model_inst: model_to_use,
+            var_map: None,
+            reverse_var_map: None,
+            solver_inst: Minisat::default(),
+            incremental_clauses: VecDeque::new(),
+        }
+    }
+
+    // Adds a new clause to the SAT solver incrementally.
+    pub fn add_incremental_clause(&mut self, clause: Vec<i32>) -> Result<(), SolverError> {
+        let new_clause: Clause = clause
+            .iter()
+            .map(|&lit| {
+                let abs_lit = lit.abs() as u32;
```

Comment:
'0' is a valid Lit value, but you can't tell from 'abs()' if it is supposed to be positive or negative.

You may have thought of this (if not, my temptation would be to just skip the '0' literal, maybe just assign it true (or false) and be done with it), but either way, might be good to add a panic here, just to handle the lit=0 case, so it's clearly covered.

Reply:
Thank you for pointing this out I have added a panic to handle lit=0